### PR TITLE
cantaloupe java opts to fix jai processor for tiffs

### PR DESCRIPTION
**Pull request by noahwsmith**
We're getting errors like this on certain TIFFs:
```
cantaloupe-1  | java.lang.IllegalAccessError: class javax.media.jai.RasterAccessor (in unnamed module @0x665bc091) cannot access class sun.awt.image.BytePackedRaster (in module java.desktop) because module java.desktop does not export sun.awt.image to unnamed module @0x665bc091
cantaloupe-1  | 	at javax.media.jai.RasterAccessor.<init>(RasterAccessor.java:802)
cantaloupe-1  | 	at com.sun.media.jai.opimage.SubsampleAverageOpImage.computeRect(SubsampleAverageOpImage.java:269)
cantaloupe-1  | 	at javax.media.jai.GeometricOpImage.computeTile(GeometricOpImage.java:683)
cantaloupe-1  | 	at com.sun.media.jai.util.SunTileScheduler.scheduleTile(SunTileScheduler.java:904)
cantaloupe-1  | 	at javax.media.jai.OpImage.getTile(OpImage.java:1129)
```

The Cantaloupe boot sequence says 
```
cantaloupe-1  | 18:06:58.351 [main] INFO  e.i.l.c.d.JRubyDelegateProxy - Loading script code
cantaloupe-1  | 2024-02-23T18:06:59.413Z [main] WARN FilenoUtil : Native subprocess control requires open access to the JDK IO subsystem
cantaloupe-1  | Pass '--add-opens java.base/sun.nio.ch=ALL-UNNAMED --add-opens java.base/java.io=ALL-UNNAMED' to enable.
```

### CODE REVIEW

**Comment by noahwsmith** [comment]
We now have this running live and formerly erroring objects are now working - e.g. 
https://compass.fivecolleges.edu/islandora/object/smith%3A771030

I didn't write an issue up for this - do we need one?

**Comment by joecorall** [comment]
@noahwsmith - what buildkit tag were you running for cantaloupe? I'm on `3.1.2` and my docker compose looks like what's below and I am able to load e.g. https://islandora-stage.lib.lehigh.edu/cantaloupe/iiif/2/https%3A%2F%2Fcompass.fivecolleges.edu%2Fsystem%2Ffiles%2F2024-01%2Fsmith_771178_OBJ.tiff/3072,2048,1024,1024/512,/0/default.jpg OK without this change.

```
$ yq '.services["cantaloupe-prod"]' docker-compose.yml
{
  "restart": "no",
  "tty": true,
  "security_opt": [
    "label=type:container_runtime_t"
  ],
  "profiles": [
    "prod"
  ],
  "secrets": [],
  "environment": {
    "CANTALOUPE_PROCESSOR_SELECTION_STRATEGY": "ManualSelectionStrategy",
    "CANTALOUPE_PROCESSOR_MANUALSELECTIONSTRATEGY_JPG": "Java2dProcessor",
    "CANTALOUPE_CACHE_SERVER_DERIVATIVE_ENABLED": "true",
    "CANTALOUPE_CACHE_SERVER_DERIVATIVE": "FilesystemCache",
  },
  "image": "${ISLANDORA_REPOSITORY}/cantaloupe:${ISLANDORA_TAG}",
  "labels": {
    "traefik.http.middlewares.https-redirect.redirectscheme.permanent": true,
    "traefik.http.middlewares.https-redirect.redirectscheme.scheme": "https",
    "traefik.enable": true,
    "traefik.http.middlewares.cantaloupe-custom-request-headers.headers.customrequestheaders.X-Forwarded-Path": "/cantaloupe",
    "traefik.http.middlewares.cantaloupe-strip-prefix.stripprefix.prefixes": "/cantaloupe",
    "traefik.http.middlewares.cantaloupe.chain.middlewares": "cantaloupe-strip-prefix,cantaloupe-custom-request-headers",
    "traefik.http.routers.cantaloupe_http.entrypoints": "http",
    "traefik.http.routers.cantaloupe_http.middlewares": "https-redirect",
    "traefik.http.routers.cantaloupe_http.rule": "Host(`${DOMAIN}`) && PathPrefix(`/cantaloupe`)",
    "traefik.http.routers.cantaloupe_http.service": "cantaloupe",
    "traefik.http.routers.cantaloupe_https.middlewares": "cantaloupe",
    "traefik.http.routers.cantaloupe_https.entrypoints": "https",
    "traefik.http.routers.cantaloupe_https.rule": "Host(`${DOMAIN}`) && PathPrefix(`/cantaloupe`)",
    "traefik.http.routers.cantaloupe_https.tls": true,
    "traefik.http.services.cantaloupe.loadbalancer.server.port": 8182,
    "traefik.http.routers.cantaloupe_https.tls.certresolver": "resolver"
  },
  "volumes": [
    "/opt/islandora/volumes/cantaloupe-data:/data:rw",
    "/opt/islandora/volumes/tmp/cantaloupe:/tmp:rw"
  ],
  "networks": {
    "default": {
      "aliases": [
        "cantaloupe"
      ]
    }
  }
}
```

**Comment by joecorall** [comment]
Hmm looks like maybe you're overriding the processor for TIFFs to use Jai. Should we be supporting Jai? Cantaloupe seems to advise against it

From https://cantaloupe-project.github.io/manual/4.0/processors.html#JaiProcessor

> [!WARNING]  
> Development on JAI ended a long time ago, and a [minor incompatibility](https://github.com/cantaloupe-project/cantaloupe/issues/176) has cropped up in Java 9. Given that supporting JAI is likely to become more problematic as time goes on, this processor should be considered deprecated, and it may be removed in a future release.

**Comment by noahwsmith** [comment]
Yes, on this project Jai is better (except this error). Jai isn't the default (which is correct) but this PR does allow Jai to work better than it does out of the box without these params. It is an edge case for sure.

**Review by joecorall — CHANGES_REQUESTED** [review]
Given Cantaloupe's docs advising against Jai, I do not think we should be adding support for the Jai processor.

I think the best path forward is to add a new environment variable `CANTALOUPE_JAVA_OPTS`. After that environment variable is created, any docker image could add Jai support for TIFFs using

```
docker run \
  --env CANTALOUPE_JAVA_OPTS="--add-opens java.base/sun.nio.ch=ALL-UNNAMED --add-opens java.base/java.io=ALL-UNNAMED --add-exports java.desktop/sun.awt.image=ALL-UNNAMED -Dcom.sun.media.jai.disableMediaLib=true" \
  islandora/cantaloupe:some-future-tag
```

**Comment by noahwsmith** [comment]
Thanks @joecorall I think that's a perfect compromise. I'll test this soon and we can get it merged this week...

**Comment by nigelgbanks** [comment]
@noahwsmith any luck with applying @joecorall's suggested changes?

**Comment by noahwsmith** [comment]
I accepted this suggested change and there are now errors on the docker build:
https://github.com/Born-Digital-US/isle-buildkit/actions/runs/8708308988/job/23885458913

**Comment by joecorall** [comment]
@noahwsmith We need to also set the env var in [the Dockerfile](https://github.com/Born-Digital-US/isle-buildkit/blob/cantaloupe-java-opts/cantaloupe/Dockerfile#L46) and update the README

**Comment by noahwsmith** [comment]
Even with those changes it does still error out
```
compass-i2-cantaloupe-1  | s6-rc: info: service s6rc-oneshot-runner: starting
compass-i2-cantaloupe-1  | s6-rc: info: service s6rc-oneshot-runner successfully started
compass-i2-cantaloupe-1  | s6-rc: info: service fix-attrs: starting
compass-i2-cantaloupe-1  | s6-rc: info: service fix-attrs successfully started
compass-i2-cantaloupe-1  | s6-rc: info: service legacy-cont-init: starting
compass-i2-cantaloupe-1  | s6-rc: info: service legacy-cont-init successfully started
compass-i2-cantaloupe-1  | s6-rc: info: service tty: starting
compass-i2-cantaloupe-1  | s6-rc: info: service container-environment: starting
compass-i2-cantaloupe-1  | s6-rc: info: service tty successfully started
compass-i2-cantaloupe-1  | s6-rc: info: service container-environment successfully started
compass-i2-cantaloupe-1  | s6-rc: info: service database-defaults: starting
compass-i2-cantaloupe-1  | s6-rc: info: service database-defaults successfully started
compass-i2-cantaloupe-1  | s6-rc: info: service confd-oneshot: starting
compass-i2-cantaloupe-1  | s6-rc: info: service confd-oneshot successfully started
compass-i2-cantaloupe-1  | s6-rc: info: service cacert-import: starting
compass-i2-cantaloupe-1  | s6-rc: info: service cacert-import successfully started
compass-i2-cantaloupe-1  | s6-rc: info: service ready: starting
compass-i2-cantaloupe-1  | s6-rc: info: service ready successfully started
compass-i2-cantaloupe-1  | s6-rc: info: service confd: starting
compass-i2-cantaloupe-1  | s6-rc: info: service cantaloupe: starting
compass-i2-cantaloupe-1  | s6-rc: info: service confd successfully started
compass-i2-cantaloupe-1  | s6-rc: info: service cantaloupe successfully started
compass-i2-cantaloupe-1  | s6-rc: info: service legacy-services: starting
compass-i2-cantaloupe-1  | s6-rc: info: service legacy-services successfully started
compass-i2-cantaloupe-1  | Error: Could not find or load main class
compass-i2-cantaloupe-1  | Caused by: java.lang.ClassNotFoundException:
compass-i2-cantaloupe-1  | s6-rc: info: service cantaloupe finish: executing...
compass-i2-cantaloupe-1  | s6-rc: info: service cantaloupe did not recieve a signal...
compass-i2-cantaloupe-1  | s6-rc: info: service cantaloupe exiting with exit code: 1
compass-i2-cantaloupe-1  | s6-rc: info: service legacy-services: stopping
compass-i2-cantaloupe-1  | s6-rc: info: service legacy-services successfully stopped
compass-i2-cantaloupe-1  | s6-rc: info: service confd: stopping
compass-i2-cantaloupe-1  | s6-rc: info: service cantaloupe: stopping
compass-i2-cantaloupe-1  | s6-rc: info: service cantaloupe successfully stopped
compass-i2-cantaloupe-1  | s6-rc: info: service confd successfully stopped
compass-i2-cantaloupe-1  | s6-rc: info: service ready: stopping
compass-i2-cantaloupe-1  | s6-rc: info: service ready successfully stopped
compass-i2-cantaloupe-1  | s6-rc: info: service tty: stopping
compass-i2-cantaloupe-1  | s6-rc: info: service cacert-import: stopping
compass-i2-cantaloupe-1  | s6-rc: info: service tty successfully stopped
compass-i2-cantaloupe-1  | s6-rc: info: service cacert-import successfully stopped
compass-i2-cantaloupe-1  | s6-rc: info: service confd-oneshot: stopping
compass-i2-cantaloupe-1  | s6-rc: info: service confd-oneshot successfully stopped
compass-i2-cantaloupe-1  | s6-rc: info: service database-defaults: stopping
compass-i2-cantaloupe-1  | s6-rc: info: service database-defaults successfully stopped
compass-i2-cantaloupe-1  | s6-rc: info: service container-environment: stopping
compass-i2-cantaloupe-1  | s6-rc: info: service container-environment successfully stopped
compass-i2-cantaloupe-1  | s6-rc: info: service legacy-cont-init: stopping
compass-i2-cantaloupe-1  | s6-rc: info: service legacy-cont-init successfully stopped
compass-i2-cantaloupe-1  | s6-rc: info: service fix-attrs: stopping
compass-i2-cantaloupe-1  | s6-rc: info: service fix-attrs successfully stopped
compass-i2-cantaloupe-1  | s6-rc: info: service s6rc-oneshot-runner: stopping
compass-i2-cantaloupe-1  | s6-rc: info: service s6rc-oneshot-runner successfully stopped
compass-i2-cantaloupe-1 exited with code 1
```

File: cantaloupe/rootfs/etc/s6-overlay/s6-rc.d/cantaloupe/run [hunk 1]
```diff
@@ -6,4 +6,4 @@ set -e
 # actually can write to the folder.
 chown cantaloupe:cantaloupe /data
 
-exec with-contenv s6-setuidgid cantaloupe java -Dcantaloupe.config=/opt/cantaloupe/cantaloupe.properties -Xms"${CANTALOUPE_HEAP_MIN}" -Xmx"${CANTALOUPE_HEAP_MAX}" -jar /opt/cantaloupe/cantaloupe.jar
+exec with-contenv s6-setuidgid cantaloupe java -Dcantaloupe.config=/opt/cantaloupe/cantaloupe.properties -Xms"${CANTALOUPE_HEAP_MIN}" -Xmx"${CANTALOUPE_HEAP_MAX}" "${CANTALOUPE_JAVA_OPTS}" -jar /opt/cantaloupe/cantaloupe.jar
```
Comment (joecorall) [inline cantaloupe/rootfs/etc/s6-overlay/s6-rc.d/cantaloupe/run:9]:
```suggestion
exec with-contenv s6-setuidgid cantaloupe java -Dcantaloupe.config=/opt/cantaloupe/cantaloupe.properties -Xms"${CANTALOUPE_HEAP_MIN}" -Xmx"${CANTALOUPE_HEAP_MAX}" ${CANTALOUPE_JAVA_OPTS} -jar /opt/cantaloupe/cantaloupe.jar
```

Comment (joecorall) [reply cantaloupe/rootfs/etc/s6-overlay/s6-rc.d/cantaloupe/run:9]: @noahwsmith - removing the double quotes around the var oughta do it

**Review by joecorall — APPROVED** [review]
```
$ make bake TARGET=cantaloupe
$ ./gradlew cantaloupe:test
Starting a Gradle Daemon (subsequent builds will be faster)

> Task :cantaloupe:tests:ServiceStartsWithDefaults:test
 Container cantaloupe-servicestartswithdefaults-cantaloupe-1  Stopping
 Container cantaloupe-servicestartswithdefaults-cantaloupe-1  Stopped

BUILD SUCCESSFUL in 15s
11 actionable tasks: 2 executed, 9 up-to-date

**Comment by noahwsmith** [comment]
Deployed and working on our client site - thanks!

**Pull request merged**